File: scripts/generate_pixel_art.py
#!/usr/bin/env python3
"""Generate the original pixel-art bounty asset for issue #80."""

from __future__ import annotations

import struct
import zlib
from pathlib import Path


ROOT = Path(__file__).resolve().parents[1]
OUT = ROOT / "assets" / "pixel-art" / "audit-lantern-grid.png"

WIDTH = 128
HEIGHT = 128

PALETTE = {
    "void": (10, 13, 20, 255),
    "deep": (16, 23, 34, 255),
    "grid": (28, 63, 70, 255),
    "trace": (62, 172, 140, 255),
    "trace_hi": (125, 238, 194, 255),
    "gold": (235, 184, 76, 255),
    "gold_hi": (255, 226, 132, 255),
    "amber": (211, 112, 57, 255),
    "shadow": (51, 36, 36, 255),
    "glass": (115, 196, 213, 255),
    "white": (235, 244, 241, 255),
}


def blank() -> list[list[tuple[int, int, int, int]]]:
    return [[PALETTE["void"] for _ in range(WIDTH)] for _ in range(HEIGHT)]


def dot(img: list[list[tuple[int, int, int, int]]], x: int, y: int, color: str) -> None:
    if 0 <= x < WIDTH and 0 <= y < HEIGHT:
        img[y][x] = PALETTE[color]


def rect(
    img: list[list[tuple[int, int, int, int]]],
    x: int,
    y: int,
    w: int,
    h: int,
    color: str,
) -> None:
    for yy in range(y, y + h):
        for xx in range(x, x + w):
            dot(img, xx, yy, color)


def line(
    img: list[list[tuple[int, int, int, int]]],
    x0: int,
    y0: int,
    x1: int,
    y1: int,
    color: str,
) -> None:
    dx = abs(x1 - x0)
    sx = 1 if x0 < x1 else -1
    dy = -abs(y1 - y0)
    sy = 1 if y0 < y1 else -1
    err = dx + dy
    while True:
        dot(img, x0, y0, color)
        if x0 == x1 and y0 == y1:
            return
        twice = 2 * err
        if twice >= dy:
            err += dy
            x0 += sx
        if twice <= dx:
            err += dx
            y0 += sy


def fill_circle(
    img: list[list[tuple[int, int, int, int]]], cx: int, cy: int, radius: int, color: str
) -> None:
    r2 = radius * radius
    for y in range(cy - radius, cy + radius + 1):
        for x in range(cx - radius, cx + radius + 1):
            if (x - cx) ** 2 + (y - cy) ** 2 <= r2:
                dot(img, x, y, color)


def pixelate_shadow(img: list[list[tuple[int, int, int, int]]]) -> None:
    for y in range(0, HEIGHT, 4):
        for x in range(0, WIDTH, 4):
            if ((x * 17 + y * 23) % 37) in {0, 1, 2}:
                rect(img, x, y, 2, 2, "deep")


def draw_grid(img: list[list[tuple[int, int, int, int]]]) -> None:
    for x in range(8, WIDTH, 12):
        line(img, x, 10, x, 116, "grid")
    for y in range(14, HEIGHT, 12):
        line(img, 8, y, 119, y, "grid")

    traces = [
        (14, 97, 45, 97, 45, 76),
        (82, 97, 111, 97, 111, 70),
        (25, 35, 48, 35, 48, 55),
        (101, 31, 82, 31, 82, 54),
        (21, 74, 37, 74, 37, 62),
        (92, 74, 109, 74, 109, 58),
    ]
    for x0, y0, x1, y1, x2, y2 in traces:
        line(img, x0, y0, x1, y1, "trace")
        line(img, x1, y1, x2, y2, "trace")
        fill_circle(img, x0, y0, 2, "trace_hi")
        fill_circle(img, x2, y2, 2, "trace_hi")


def draw_lantern(img: list[list[tuple[int, int, int, int]]]) -> None:
    fill_circle(img, 64, 62, 29, "amber")
    fill_circle(img, 64, 62, 23, "gold")
    fill_circle(img, 64, 62, 15, "gold_hi")

    rect(img, 45, 42, 38, 48, "shadow")
    rect(img, 49, 46, 30, 40, "glass")
    rect(img, 53, 50, 22, 32, "deep")
    rect(img, 57, 54, 14, 24, "gold_hi")
    rect(img, 60, 58, 8, 16, "white")

    rect(img, 42, 39, 44, 5, "gold")
    rect(img, 47, 35, 34, 4, "gold_hi")
    rect(img, 55, 30, 18, 5, "gold")
    line(img, 58, 30, 58, 22, "gold_hi")
    line(img, 70, 30, 70, 22, "gold_hi")
    line(img, 58, 22, 70, 22, "gold")

    rect(img, 42, 90, 44, 6, "gold")
    rect(img, 49, 96, 30, 4, "gold_hi")
    rect(img, 55, 100, 18, 4, "gold")

    line(img, 50, 46, 78, 82, "trace_hi")
    line(img, 78, 46, 50, 82, "trace_hi")
    for x, y in [(52, 49), (75, 49), (52, 79), (75, 79)]:
        fill_circle(img, x, y, 2, "white")


def draw_foreground_marks(img: list[list[tuple[int, int, int, int]]]) -> None:
    for x, y in [(20, 106), (33, 111), (94, 110), (108, 104), (17, 25), (111, 24)]:
        rect(img, x, y, 5, 5, "gold")
        rect(img, x + 1, y + 1, 3, 3, "gold_hi")
    for x, y in [(30, 84), (98, 84), (64, 111), (64, 18)]:
        fill_circle(img, x, y, 3, "trace_hi")
        fill_circle(img, x, y, 1, "white")


def write_png(path: Path, img: list[list[tuple[int, int, int, int]]]) -> None:
    raw = b"".join(b"\x00" + b"".join(bytes(pixel) for pixel in row) for row in img)

    def chunk(kind: bytes, data: bytes) -> bytes:
        return (
            struct.pack(">I", len(data))
            + kind
            + data
            + struct.pack(">I", zlib.crc32(kind + data) & 0xFFFFFFFF)
        )

    path.parent.mkdir(parents=True, exist_ok=True)
    path.write_bytes(
        b"\x89PNG\r\n\x1a\n"
        + chunk(b"IHDR", struct.pack(">IIBBBBB", WIDTH, HEIGHT, 8, 6, 0, 0, 0))
        + chunk(b"IDAT", zlib.compress(raw, 9))
        + chunk(b"IEND", b"")
    )


def main() -> None:
    img = blank()
    pixelate_shadow(img)
    draw_grid(img)
    draw_lantern(img)
    draw_foreground_marks(img)
    write_png(OUT, img)
    print(OUT)


if __name__ == "__main__":
    main()

File: scripts/generate-pixel-art.mjs
import { writeFileSync } from "node:fs";
import { deflateSync } from "node:zlib";

const width = 128;
const height = 128;
const out = "assets/pixel-art/proof-gate-console.png";

const palette = {
  bg0: [15, 23, 42, 255],
  bg1: [30, 41, 59, 255],
  grid: [51, 65, 85, 255],
  frame: [71, 85, 105, 255],
  frameLight: [148, 163, 184, 255],
  panel: [17, 24, 39, 255],
  panel2: [30, 64, 175, 255],
  teal: [45, 212, 191, 255],
  cyan: [34, 211, 238, 255],
  green: [74, 222, 128, 255],
  amber: [251, 191, 36, 255],
  orange: [249, 115, 22, 255],
  red: [248, 113, 113, 255],
  white: [226, 232, 240, 255],
  banana: [250, 204, 21, 255],
  bananaDark: [180, 83, 9, 255],
  shadow: [2, 6, 23, 255]
};

const pixels = new Uint8Array(width * height * 4);

function put(x, y, color) {
  if (x < 0 || y < 0 || x >= width || y >= height) return;
  pixels.set(color, (y * width + x) * 4);
}

function rect(x, y, w, h, color) {
  for (let yy = y; yy < y + h; yy++) {
    for (let xx = x; xx < x + w; xx++) put(xx, yy, color);
  }
}

function line(x0, y0, x1, y1, color) {
  const dx = Math.abs(x1 - x0);
  const sx = x0 < x1 ? 1 : -1;
  const dy = -Math.abs(y1 - y0);
  const sy = y0 < y1 ? 1 : -1;
  let err = dx + dy;
  let x = x0;
  let y = y0;
  while (true) {
    put(x, y, color);
    if (x === x1 && y === y1) break;
    const e2 = 2 * err;
    if (e2 >= dy) {
      err += dy;
      x += sx;
    }
    if (e2 <= dx) {
      err += dx;
      y += sy;
    }
  }
}

function crc32(buffer) {
  let crc = ~0;
  for (const byte of buffer) {
    crc ^= byte;
    for (let i = 0; i < 8; i++) {
      crc = (crc >>> 1) ^ (0xedb88320 & -(crc & 1));
    }
  }
  return ~crc >>> 0;
}

function chunk(type, data) {
  const typeBytes = Buffer.from(type);
  const length = Buffer.alloc(4);
  length.writeUInt32BE(data.length);
  const crc = Buffer.alloc(4);
  crc.writeUInt32BE(crc32(Buffer.concat([typeBytes, data])));
  return Buffer.concat([length, typeBytes, data, crc]);
}

function writePng() {
  const scanlines = Buffer.alloc((width * 4 + 1) * height);
  for (let y = 0; y < height; y++) {
    const rowStart = y * (width * 4 + 1);
    scanlines[rowStart] = 0;
    for (let x = 0; x < width * 4; x++) {
      scanlines[rowStart + 1 + x] = pixels[y * width * 4 + x];
    }
  }

  const ihdr = Buffer.alloc(13);
  ihdr.writeUInt32BE(width, 0);
  ihdr.writeUInt32BE(height, 4);
  ihdr[8] = 8;
  ihdr[9] = 6;
  ihdr[10] = 0;
  ihdr[11] = 0;
  ihdr[12] = 0;

  writeFileSync(out, Buffer.concat([
    Buffer.from([137, 80, 78, 71, 13, 10, 26, 10]),
    chunk("IHDR", ihdr),
    chunk("IDAT", deflateSync(scanlines, { level: 9 })),
    chunk("IEND", Buffer.alloc(0))
  ]));
}

rect(0, 0, width, height, palette.bg0);
for (let y = 0; y < height; y += 8) line(0, y, width - 1, y, palette.bg1);
for (let x = 0; x < width; x += 8) line(x, 0, x, height - 1, palette.bg1);

rect(12, 84, 104, 20, palette.shadow);
rect(16, 18, 96, 78, palette.frame);
rect(20, 22, 88, 70, palette.panel);
rect(24, 26, 80, 46, palette.bg0);
rect(28, 30, 72, 38, palette.panel2);

rect(32, 36, 16, 10, palette.cyan);
rect(56, 36, 16, 10, palette.green);
rect(80, 36, 16, 10, palette.amber);
rect(32, 54, 16, 6, palette.green);
rect(56, 54, 16, 6, palette.cyan);
rect(80, 54, 16, 6, palette.orange);

line(48, 41, 56, 41, palette.white);
line(72, 41, 80, 41, palette.white);
line(48, 57, 56, 57, palette.white);
line(72, 57, 80, 57, palette.white);

rect(30, 76, 68, 10, palette.frameLight);
rect(34, 78, 8, 4, palette.teal);
rect(46, 78, 8, 4, palette.green);
rect(58, 78, 8, 4, palette.amber);
rect(70, 78, 8, 4, palette.orange);
rect(82, 78, 8, 4, palette.red);

rect(45, 94, 38, 10, palette.frame);
rect(38, 104, 52, 8, palette.frameLight);
rect(34, 112, 60, 4, palette.shadow);

rect(50, 102, 28, 3, palette.bananaDark);
rect(48, 98, 6, 6, palette.banana);
rect(54, 96, 18, 8, palette.banana);
rect(72, 98, 6, 6, palette.banana);
rect(56, 100, 14, 2, palette.white);

line(24, 116, 104, 116, palette.teal);
line(24, 118, 104, 118, palette.cyan);
rect(22, 112, 8, 8, palette.green);
rect(50, 112, 8, 8, palette.amber);
rect(78, 112, 8, 8, palette.green);
rect(100, 112, 8, 8, palette.amber);

writePng();
